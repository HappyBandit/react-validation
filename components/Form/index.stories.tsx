import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';
import { Meta } from '@storybook/react';
import { useInputValidation } from '../hooks';
import { email, phone, required } from '../Validators';
import FormComponent from './index';

const Text = ({ value, onChange }: { value: string; onChange: (value: string) => void }) => {
    const result = useInputValidation([ required(value, 'ENTER SOMETHING') ]);

    return (
        <div className="col-4 mb-3">
            <label htmlFor="validationCustomUsername">Username*</label>
            <div className="input-group">
                <div className="input-group-prepend">
                    <span className="input-group-text" id="inputGroupPrepend">@</span>
                </div>
                <input
                    type="text"
                    className={`form-control ${result.valid ? '' : 'is-invalid'}`}
                    id="validationCustomUsername"
                    placeholder="Username"
                    aria-describedby="inputGroupPrepend"
                    onChange={(event) => onChange(event.currentTarget.value)}
                />
                <div className="invalid-feedback">
                    {result.message}
                </div>
            </div>
        </div>
    );
};

const Phone = ({ value, onChange }: { value: string; onChange: (value: string) => void }) => {
    const result = useInputValidation([ phone(value, 'ENTER A PHONE') ]);

    return (
        <div className="col-4 mb-3">
            <label htmlFor="phone">Phone Number</label>
            <input
                type="text"
                className={`form-control ${result.valid ? '' : 'is-invalid'}`}
                id="phone"
                placeholder="Phone Number"
                onChange={(event) => onChange(event.currentTarget.value)}
            />
            <div className="invalid-feedback">
                {result.message}
            </div>
        </div>
    );
};

const Email = ({ value, onChange }: { value: string; onChange: (value: string) => void }) => {
    const result = useInputValidation([
        required(value, 'ENTER SOMETHING'),
        email(value, 'MUST BE A VALID EMAIL'),
    ]);

    return (
        <div className="col-4 mb-3">
            <label htmlFor="email">Email*</label>
            <input
                type="text"
                className={`form-control ${result.valid ? '' : 'is-invalid'}`}
                id="email"
                placeholder="Email"
                onChange={(event) => onChange(event.currentTarget.value)}
            />
            <div className="invalid-feedback">
                {result.message}
            </div>
        </div>
    );
};

const ValidationStory = () => {
    const [ text, setText ] = useState(undefined);
    const [ phone, setPhone ] = useState(undefined);
    const [ email, setEmail ] = useState(undefined);

    const handleSubmit = (...args) => {
        action('submit')(...args, text, phone, email);
    };
    return (
        <FormComponent
            onSubmit={handleSubmit}
        >
            <div className="row">
                <Text value={text} onChange={(value) => setText(value)}/>
                <Phone value={phone} onChange={(value) => setPhone(value)}/>
                <Email value={email} onChange={(value) => setEmail(value)}/>
            </div>
            <button className="btn btn-primary">Submit</button>
        </FormComponent>
    );
};

export default {
    title: 'Components/Form',
    component: FormComponent,
} as Meta;

export const Form = () => <ValidationStory/>;
Form.parameters = {
    controls: { hideNoControlsWarning: true },
};
