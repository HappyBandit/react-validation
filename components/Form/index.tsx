import React, { useRef } from 'react';
import FormContext from '../formContext';
import { IFormProps, IInputObject } from '../interfaces';

const Form: React.FunctionComponent<IFormProps> = ( { onSubmit, ...formProps } ): React.ReactElement => {
    const { current: inputs } = useRef<{ [key: string]: IInputObject }>( {} );

    const addInput = ( uuid, newInput: IInputObject ): void => {
        inputs[uuid] = newInput;
    };
    const handleSubmit = ( event: React.FormEvent ) => {
        event.preventDefault();

        let valid = true;

        Object.values( inputs ).forEach( ( input ) => {
            if ( !input.isValid() ) {
                valid = false;
            }
        } );

        onSubmit( valid );
    };

    const context = {
        register: addInput,
    };

    return (
        <FormContext.Provider value={context}>
            <form {...formProps} onSubmit={handleSubmit}/>
        </FormContext.Provider>
    );
};

export default Form;
