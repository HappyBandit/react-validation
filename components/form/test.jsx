import React from 'react';
import { shallow } from 'enzyme';

import Form from './';

describe('Form', () => {
    it('should render', () => {
        const tree = shallow(
            <Form className="class">
                Child
            </Form>
        );

        expect(tree).toMatchSnapshot();
    });

    it('should return true when all validators return true', () => {
        const tree = shallow(
            <Form className="class">
                Child
            </Form>
        );

        tree.instance().registerValidator({
            isValid: () => true
        });

        tree.instance().registerValidator({
            isValid: () => true
        });

        expect(tree.instance().isValid()).toEqual(true);
    });

    it('should return false when any validators return false', () => {
        const tree = shallow(
            <Form className="class">
                Child
            </Form>
        );

        tree.instance().registerValidator({
            isValid: () => true
        });

        tree.instance().registerValidator({
            isValid: () => false
        });

        expect(tree.instance().isValid()).toEqual(false);
    });

    it('should add validator when registerValidator called', () => {
        const tree = shallow(
            <Form className="class">
                Child
            </Form>
        );

        tree.instance().registerValidator('validator1');
        tree.instance().registerValidator('validator2');
        tree.instance().registerValidator('validator3');

        expect(tree.state('validators')).toEqual(['validator1', 'validator2', 'validator3']);
    });

    it('should return isValid response to onSubmit prop', () => {
        const mockCallback = jest.fn();
        const mockEvent = {
            preventDefault: jest.fn()
        };

        const tree = shallow(
            <Form className="class" onSubmit={mockCallback}>
                Child
            </Form>
        );

        tree.instance().isValid = () => 'VALID';
        tree.instance().handleSubmit(mockEvent);

        expect(mockCallback.mock.calls[0][0]).toEqual('VALID');
        expect(mockEvent.preventDefault.mock.calls.length).toBe(1);
    });
});
