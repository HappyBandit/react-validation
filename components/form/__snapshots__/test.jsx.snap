// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render 1`] = `
<[object Object]
  value={
    Object {
      "errorComponent": [Function],
      "register": [Function],
    }
  }
>
  <form
    className="class"
    onSubmit={[Function]}
  >
    Child
  </form>
</[object Object]>
`;
